server/resources/GmailResources.py

```diff
@@ -0,0 +1,159 @@
+from apiclient.discovery import build
+from oauth2client.client import FlowExchangeError
+from oauth2client.client import OAuth2Credentials
+from oauth2client.client import flow_from_clientsecrets
+from flask_restful import Resource, reqparse
+from flask_jwt_extended import jwt_required, get_jwt_identity
+import logging
+import httplib2
+import pickle
+
+from models.GmailCredentials import GmailCredentials
+from models.UserModel import UserModel
+
+authorize_parser = reqparse.RequestParser()
+authorize_parser.add_argument('code', required=True)
+
+CLIENTSECRETS_LOCATION = 'instance/client_secrets.json'
+REDIRECT_URI = 'http://localhost:5000/'
+SCOPES = [
+    'https://www.googleapis.com/auth/gmail.readonly',
+    'https://www.googleapis.com/auth/userinfo.email',
+    'https://www.googleapis.com/auth/userinfo.profile',
+    'https://www.googleapis.com/auth/gmail.send'
+]
+
+
+class GetCredentialsException(Exception):
+    """Error raised when an error occurred while retrieving credentials."""
+
+
+class CodeExchangeException(GetCredentialsException):
+    """Error raised when a code exchange has failed."""
+
+
+class NoRefreshTokenException(GetCredentialsException):
+    """Error raised when no refresh token has been found."""
+
+
+class NoUserIdException(Exception):
+    """Error raised when no user ID could be retrieved."""
+
+
+def get_stored_credentials(user_id):
+    """Retrieve stored credentials for the provided user ID. 
+    (app id, not Gmail id)
+
+    Args:
+      user_id: User's ID.
+    Returns:
+      Stored oauth2client.client.OAuth2Credentials if found, None otherwise.
+    """
+    query_result = GmailCredentials.filter_by(id=user_id).first()
+    if query_result is None:
+        return None
+    credentials = OAuth2Credentials.from_json(query_result.gmail_credentials)
+    if credentials.access_token_expired():
+        credentials.refresh(httplib2.Http())
+        # update db with refreshed token
+        query_result.gmail_credentials = credentials.to_json()
```

Comment:
We were told that we don't need to be using refresh token for now, hence why we made access token to last 30 days.

Reply:
Our JWT tokens are supposed to be good for 30 days, but the google oauth2 credentials expire after 60 minutes, so we'll need to refresh those.